### PR TITLE
Include callbacks

### PR DESCRIPTION
**Pull request by RoyStegeman**
I had to include the processing of callbacks. 

Also, we used `get_number_nodes` to get the number of nodes in each layer, currently the pdf model is considered a single layer and the output nodes are  always 14, so I went back to relying on the weight-bias structure which we still use later on anyway. 

The conda package should be updated as well, this is what causes CI to fail at https://github.com/NNPDF/nnpdf/pull/737.

### CODE REVIEW

**Comment by scarlehoff** [comment]
wrt the callbacks they need to be implemented also on epoch begin (and in some other situations) to ensure they work like tf: https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback

> Also, we used `get_number_nodes` to get the number of nodes in each layer, currently the pdf model is considered a single layer and the output nodes are always 14, so I went back to relying on the weight-bias structure which we still use later on anyway.

Well, it is not a layer but rather a model so the correct thing is to have a function that does:

```python
def get_nodes_model(model):
     for layer in model.layers:
          if isinstance(layer, Model):
              nodes += get_nodes_model(layer)
          elif hasattr(layer, 'bias'):
              nodes += layer.bias.shape[0]
          else:
              raise ValueError("bad boy")
```
For instance.

If we only accept weight-bias structure I think we need to check for it and fail if it fails. Keras gives you a `bias` attribute so we can use that.

Btw, rather than commenting the function I would move the functionality to that function so that you can still have a test that it is doing what you think it is doing. And add there the check (be it a raise Exception or a logger.Warning).

In any case, maybe it is a good point to either take care of the TODO and use the layers themselves or get rid of the TO-DOs, embrace the weight-bias structure but in that case you need to fail early as soon as you find something that does not follow the given structure.

**Comment by RoyStegeman** [comment]
Oh sorry! I should've made this PR a draft, as I wasn't quite finished yet, I just got it to work again with n3fit. 

Anyway, yes, the callback options at other points should also be included. 

> If we only accept weight-bias structure I think we need to check for it and fail if it fails. Keras gives you a bias attribute so we can use that.

Thanks for pointing out the excistence of the 'bias' attribute. I think this is what we should do. The NGA is specifically designed to train NN's with a weight-bias structure, so I don't see the harm in enforcing this. The alternative would be to either not train parameters that don't adhere to this structure, or to train those as if it were a regular GA, which I think kind of forgoes the point of the algorithm (and also it doesn't really seem worth the effort since it's likely never going to be used). 

I'll try to do this at some point this week.

**Comment by scarlehoff** [comment]
No worry, there's no rush, I am going on holiday tomorrow (and the plenary talks of ICHEP start in one hour) so wanted to have a look "before leaving" :P

**Comment by RoyStegeman** [comment]
Oh I see. When will you be back? Just so I'll know not to bother you with questions ;). 

And enjoy your holiday!

**Comment by scarlehoff** [comment]
The 15th, but I'll be alive if you need something :P

**Comment by RoyStegeman** [comment]
Hi @scarlehoff, I think this'll have to do regarding the weight-bias check. Would you mind checking whether you agree?

File: src/evolutionary_keras/models.py [hunk 1]
```diff
@@ -98,23 +101,55 @@ def perform_genetic_fit(
                 information = f" > epoch: {epoch+1}/{epochs}, {loss} "
                 log.info(information)
 
+            epoch_logs.update(score)
```
Comment (scarlehoff) [inline src/evolutionary_keras/models.py:104]: Why the use of update here? As far as I can see you get the same results using `score` instead of `epoch_logs` all the way through. Maybe I'm missing something.

File: src/evolutionary_keras/models.py [hunk 2]
```diff
@@ -98,23 +101,55 @@ def perform_genetic_fit(
...
             self.history_info.on_epoch_end(epoch, history_data)
 
+        callbacks.on_train_end(logs=epoch_logs)
+
         return self.history_info
 
-    def fit(self, x=None, y=None, validation_data=None, epochs=1, verbose=0, **kwargs):
+    def genetic_fit(
+        self, x=None, y=None, validation_data=None, epochs=1, verbose=0, callbacks=None,
+    ):
```
Comment (scarlehoff) [inline src/evolutionary_keras/models.py:117]: some docs here

File: src/evolutionary_keras/optimizers.py [hunk 3]
```diff
@@ -114,18 +105,29 @@ def get_shape(self):
...
+                    NNshape = "incompatible"
+            else:
+                if np.array(layer_shape).size != 2:
+                    NNshape = "incompatible"
+        if count_nodes != self.n_nodes:
+            NNshape = "incompatible"
+        if NNshape == "incompatible":
+            raise ValueError(
+                "Using the NGA optimizer requires the network to have a (weight-bias)^n structure, with 1-dimensional bias"
+            )
```
Comment (scarlehoff) [inline src/evolutionary_keras/optimizers.py:130]: Rather than setting having a `NNshape` variable as a string, I would return the exception at every point, defining exactly why the exception happened in each case.

File: src/evolutionary_keras/models.py (hunk 1 above)
Comment (RoyStegeman) [reply src/evolutionary_keras/models.py:104]: To be honest, I just looked at what is done in tf: https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/engine/training.py#L824-L1146

But I expect that the the reason loss-history is also stored in the callback logs, is in case it is needed for something (I'm thinking of e.g. stopping). I'm not sure what kind of use the the callback at the end of the training can have, but also there I'd imagine the history could be of importance rather than just the final epoch (e.g. in case callback is interested in the best result).

Comment (scarlehoff) [reply src/evolutionary_keras/models.py:104]: I see, from what I see in that file the loop is over the possible output layers of the model, so you want to update so that the dictionary stores a reference to each output.
In this case you are looping over the number of epochs so the score dictionary will always have the same entries (so update is just replacing the full dictionary each time)

**Comment by RoyStegeman** [comment]
Thanks for the comments, I've adressed them. 

> Does it also work already with n3fit? 

It did two weeks ago, and it's currently running a test after I rebased. But we'd have to update the conda package to prevent pytest of nnpdf failing. 

Also, back in April I didn't really see the harm in including this in nnpdf, but if it's not going to be used anymore (with the exception of **maybe** very rarely for some test), I might prefer not to clutter n3fit with it. Or since I believe both you and Stefano are in favor of including this, why is that? Are there already plans for using it in the future?

**Comment by scarlehoff** [comment]
> Or since I believe both you and Stefano are in favor of including this, why is that? Are there already plans for using it in the future?

tbh I have no strong feelings one way or another

**Comment by scarlehoff** [comment]
Also the conda package will in theory autoupdate itself once the new release is drafted.
